### PR TITLE
Support for HashIndexes in Orca

### PR DESCRIPTION
**Pull request by pobbatihari**
Currently ORCA is not using hash indexes while generating plan. This PR 
allows to generate plan with hashindexes in ORCA. 

- Push down the hashindex type to dxl
- Use hash indexes at CXformSelect2BitmapBoolOp, ExfSelect2IndexGet, ExfSelect2DynamicIndexGet, 
    ExfSelect2DynamicBitmapBoolOp to generate plane
- Generate a plan with indexscan by using hash indexes when bitmapscan was turned off as like planner.

setup:
create table foo(a int, b int);
create table bar(a int, b int);
insert into foo select i, i from generate_series(1, 1000000)i;
insert into bar select i, i from generate_series(1, 1000000)i;
analyze foo;
analyze bar;
create index idx on foo using hash(b);

Before Fix:
```
# explain analyze select * from foo where b=2;
                                                         QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..448.93 rows=1 width=8) (actual time=126.087..159.131 rows=1 loops=1)
   ->  Seq Scan on foo  (cost=0.00..448.93 rows=1 width=8) (actual time=2.566..121.557 rows=1 loops=1)
         Filter: (b = 2)
         Rows Removed by Filter: 333166
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 48.875 ms
   (slice0)    Executor memory: 36K bytes.
   (slice1)    Executor memory: 37K bytes avg x 3 workers, 37K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 194.206 ms
(10 rows)

# explain analyze select * from foo where b=2 or b=3;
                                                         QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..448.93 rows=2 width=8) (actual time=189.272..189.597 rows=2 loops=1)
   ->  Seq Scan on foo  (cost=0.00..448.93 rows=1 width=8) (actual time=0.279..188.945 rows=2 loops=1)
         Filter: ((b = 2) OR (b = 3))
         Rows Removed by Filter: 333165
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 7.450 ms
   (slice0)    Executor memory: 36K bytes.
   (slice1)    Executor memory: 36K bytes avg x 3 workers, 36K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 190.123 ms
(10 rows)

# explain analyze select * from foo where b=2 and a=2;
                                                         QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..459.90 rows=1 width=8) (actual time=140.956..140.958 rows=1 loops=1)
   ->  Seq Scan on foo  (cost=0.00..459.90 rows=1 width=8) (actual time=0.231..140.679 rows=1 loops=1)
         Filter: ((b = 2) AND (a = 2))
         Rows Removed by Filter: 333166
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 5.606 ms
   (slice0)    Executor memory: 36K bytes.
   (slice1)    Executor memory: 36K bytes (seg0).
 Memory used:  128000kB
 Execution Time: 141.472 ms
(10 rows)

# explain analyze select * from foo, bar where foo.b = bar.b and foo.b=2;
                                                                   QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..897.87 rows=1 width=16) (actual time=258.924..260.695 rows=1 loops=1)
   ->  Hash Join  (cost=0.00..897.87 rows=1 width=16) (actual time=158.211..258.463 rows=1 loops=1)
         Hash Cond: (foo.b = bar.b)
         Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 1 of 524288 buckets.
         ->  Seq Scan on foo  (cost=0.00..448.93 rows=1 width=8) (actual time=0.312..99.112 rows=1 loops=1)
               Filter: (b = 2)
               Rows Removed by Filter: 333166
         ->  Hash  (cost=448.93..448.93 rows=1 width=8) (actual time=154.567..154.568 rows=1 loops=1)
               Buckets: 524288  Batches: 1  Memory Usage: 4097kB
               ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..448.93 rows=1 width=8) (actual time=154.393..154.547 rows=1 loops=1)
                     ->  Seq Scan on bar  (cost=0.00..448.93 rows=1 width=8) (actual time=2.322..155.525 rows=1 loops=1)
                           Filter: (b = 2)
                           Rows Removed by Filter: 333166
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 26.376 ms
   (slice0)    Executor memory: 45K bytes.
   (slice1)    Executor memory: 4151K bytes avg x 3 workers, 4151K bytes max (seg0).  Work_mem: 4097K bytes max.
   (slice2)    Executor memory: 37K bytes avg x 3 workers, 37K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 288.814 ms
(20 rows)
```

AfterFix:

```
# explain analyze select * from foo where b=2;
                                                      QUERY PLAN
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..6.00 rows=1 width=8) (actual time=0.525..0.528 rows=1 loops=1)
   ->  Index Scan using idx on foo  (cost=0.00..6.00 rows=1 width=8) (actual time=0.016..0.017 rows=1 loops=1)
         Index Cond: (b = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 9.585 ms
   (slice0)    Executor memory: 39K bytes.
   (slice1)    Executor memory: 39K bytes avg x 3 workers, 39K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 1.193 ms
(9 rows)

# explain analyze select * from foo where b=2 or b=3;
                                                       QUERY PLAN
------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.06 rows=3 width=8) (actual time=0.707..0.963 rows=2 loops=1)
   ->  Bitmap Heap Scan on foo  (cost=0.00..431.06 rows=1 width=8) (actual time=0.030..0.040 rows=2 loops=1)
         Recheck Cond: ((b = 2) OR (b = 3))
         ->  BitmapOr  (cost=0.00..0.00 rows=0 width=0) (actual time=0.018..0.019 rows=1 loops=1)
               ->  Bitmap Index Scan on idx  (cost=0.00..0.00 rows=0 width=0) (actual time=0.010..0.011 rows=1 loops=1)
                     Index Cond: (b = 2)
               ->  Bitmap Index Scan on idx  (cost=0.00..0.00 rows=0 width=0) (actual time=0.004..0.004 rows=1 loops=1)
                     Index Cond: (b = 3)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 6.825 ms
   (slice0)    Executor memory: 51K bytes.
   (slice1)    Executor memory: 176K bytes avg x 3 workers, 176K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 1.996 ms
(14 rows)

# explain analyze select * from foo where b=2 and a=2;
                                                      QUERY PLAN
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..6.00 rows=1 width=8) (actual time=0.405..0.406 rows=1 loops=1)
   ->  Index Scan using idx on foo  (cost=0.00..6.00 rows=1 width=8) (actual time=0.018..0.019 rows=1 loops=1)
         Index Cond: (b = 2)
         Filter: (a = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 8.937 ms
   (slice0)    Executor memory: 40K bytes.
   (slice1)    Executor memory: 40K bytes (seg0).
 Memory used:  128000kB
 Execution Time: 1.026 ms
(10 rows)

# explain analyze select * from foo, bar where foo.a = bar.a and foo.b=2;
                                                         QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..508.48 rows=1 width=16) (actual time=118.941..118.943 rows=1 loops=1)
   ->  Hash Join  (cost=0.00..508.48 rows=1 width=16) (actual time=0.691..118.413 rows=1 loops=1)
         Hash Cond: (bar.a = foo.a)
         Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 1 of 524288 buckets.
         ->  Seq Scan on bar  (cost=0.00..437.97 rows=333334 width=8) (actual time=0.276..50.627 rows=333167 loops=1)
         ->  Hash  (cost=6.00..6.00 rows=1 width=8) (actual time=0.027..0.028 rows=1 loops=1)
               Buckets: 524288  Batches: 1  Memory Usage: 4097kB
               ->  Index Scan using idx on foo  (cost=0.00..6.00 rows=1 width=8) (actual time=0.023..0.024 rows=1 loops=1)
                     Index Cond: (b = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 16.278 ms
   (slice0)    Executor memory: 43K bytes.
   (slice1)    Executor memory: 4139K bytes avg x 3 workers, 4163K bytes max (seg0).  Work_mem: 4097K bytes max.
 Memory used:  128000kB
 Execution Time: 119.748 ms
(15 rows)
```

set optimizer_enable_bitmapscan to on;
set optimizer_enable_indexscan to off;

```
# explain analyze select * from foo where b=2;
                                                       QUERY PLAN
------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..391.30 rows=1 width=8) (actual time=0.774..0.800 rows=1 loops=1)
   ->  Bitmap Heap Scan on foo  (cost=0.00..391.30 rows=1 width=8) (actual time=0.024..0.033 rows=1 loops=1)
         Recheck Cond: (b = 2)
         ->  Bitmap Index Scan on idx  (cost=0.00..0.00 rows=0 width=0) (actual time=0.012..0.012 rows=1 loops=1)
               Index Cond: (b = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 6.649 ms
   (slice0)    Executor memory: 43K bytes.
   (slice1)    Executor memory: 161K bytes avg x 3 workers, 161K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 2.018 ms
(11 rows)

# explain analyze select * from foo where b=2 or b=3;
                                                       QUERY PLAN
------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.06 rows=3 width=8) (actual time=0.665..0.667 rows=2 loops=1)
   ->  Bitmap Heap Scan on foo  (cost=0.00..431.06 rows=1 width=8) (actual time=0.045..0.065 rows=2 loops=1)
         Recheck Cond: ((b = 2) OR (b = 3))
         ->  BitmapOr  (cost=0.00..0.00 rows=0 width=0) (actual time=0.032..0.033 rows=1 loops=1)
               ->  Bitmap Index Scan on idx  (cost=0.00..0.00 rows=0 width=0) (actual time=0.021..0.021 rows=1 loops=1)
                     Index Cond: (b = 2)
               ->  Bitmap Index Scan on idx  (cost=0.00..0.00 rows=0 width=0) (actual time=0.007..0.007 rows=1 loops=1)
                     Index Cond: (b = 3)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 7.365 ms
   (slice0)    Executor memory: 51K bytes.
   (slice1)    Executor memory: 176K bytes avg x 3 workers, 176K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 1.835 ms
(14 rows)

# explain analyze select * from foo where b=2 and a=2;
                                                       QUERY PLAN
------------------------------------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..397.97 rows=1 width=8) (actual time=0.589..0.590 rows=1 loops=1)
   ->  Bitmap Heap Scan on foo  (cost=0.00..397.97 rows=1 width=8) (actual time=0.023..0.042 rows=1 loops=1)
         Recheck Cond: (b = 2)
         Filter: (a = 2)
         ->  Bitmap Index Scan on idx  (cost=0.00..0.00 rows=0 width=0) (actual time=0.010..0.011 rows=1 loops=1)
               Index Cond: (b = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 8.150 ms
   (slice0)    Executor memory: 44K bytes.
   (slice1)    Executor memory: 162K bytes (seg0).
 Memory used:  128000kB
 Execution Time: 1.505 ms
(12 rows)

# explain analyze select * from foo, bar where foo.a = bar.a and foo.b=2;
                                                          QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..900.45 rows=1 width=16) (actual time=179.010..179.013 rows=1 loops=1)
   ->  Hash Join  (cost=0.00..900.45 rows=1 width=16) (actual time=0.981..178.302 rows=1 loops=1)
         Hash Cond: (bar.a = foo.a)
         Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 1 of 262144 buckets.
         ->  Seq Scan on bar  (cost=0.00..437.97 rows=333334 width=8) (actual time=0.187..57.367 rows=333167 loops=1)
         ->  Hash  (cost=397.97..397.97 rows=1 width=8) (actual time=0.105..0.106 rows=1 loops=1)
               Buckets: 262144  Batches: 1  Memory Usage: 2049kB
               ->  Bitmap Heap Scan on foo  (cost=0.00..397.97 rows=1 width=8) (actual time=0.028..0.103 rows=1 loops=1)
                     Recheck Cond: (b = 2)
                     ->  Bitmap Index Scan on idx  (cost=0.00..0.00 rows=0 width=0) (actual time=0.011..0.011 rows=1 loops=1)
                           Index Cond: (b = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 16.392 ms
   (slice0)    Executor memory: 61K bytes.
   (slice1)    Executor memory: 2231K bytes avg x 3 workers, 2252K bytes max (seg0).  Work_mem: 2049K bytes max.
 Memory used:  128000kB
 Execution Time: 180.051 ms
(17 rows)
```

set optimizer_enable_bitmapscan to off;
set optimizer_enable_indexscan to on;

```
# explain analyze select * from foo where b=2;
                                                      QUERY PLAN
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..6.00 rows=1 width=8) (actual time=0.387..0.421 rows=1 loops=1)
   ->  Index Scan using idx on foo  (cost=0.00..6.00 rows=1 width=8) (actual time=0.020..0.021 rows=1 loops=1)
         Index Cond: (b = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 5.571 ms
   (slice0)    Executor memory: 39K bytes.
   (slice1)    Executor memory: 38K bytes avg x 3 workers, 38K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 1.107 ms
(9 rows)

# explain analyze select * from foo where b=2 or b=3;
                                                         QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..448.93 rows=3 width=8) (actual time=116.449..120.748 rows=2 loops=1)
   ->  Seq Scan on foo  (cost=0.00..448.93 rows=1 width=8) (actual time=0.296..116.166 rows=2 loops=1)
         Filter: ((b = 2) OR (b = 3))
         Rows Removed by Filter: 333165
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 5.608 ms
   (slice0)    Executor memory: 36K bytes.
   (slice1)    Executor memory: 36K bytes avg x 3 workers, 36K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 121.354 ms
(10 rows)

# explain analyze select * from foo where b=2 and a=2;
                                                      QUERY PLAN
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..6.00 rows=1 width=8) (actual time=0.411..0.413 rows=1 loops=1)
   ->  Index Scan using idx on foo  (cost=0.00..6.00 rows=1 width=8) (actual time=0.019..0.020 rows=1 loops=1)
         Index Cond: (b = 2)
         Filter: (a = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 6.894 ms
   (slice0)    Executor memory: 40K bytes.
   (slice1)    Executor memory: 40K bytes (seg0).
 Memory used:  128000kB
 Execution Time: 1.029 ms
(10 rows)

# explain analyze select * from foo, bar where foo.a = bar.a and foo.b=2;
                                                         QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..508.48 rows=1 width=16) (actual time=118.941..118.943 rows=1 loops=1)
   ->  Hash Join  (cost=0.00..508.48 rows=1 width=16) (actual time=0.691..118.413 rows=1 loops=1)
         Hash Cond: (bar.a = foo.a)
         Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 1 of 524288 buckets.
         ->  Seq Scan on bar  (cost=0.00..437.97 rows=333334 width=8) (actual time=0.276..50.627 rows=333167 loops=1)
         ->  Hash  (cost=6.00..6.00 rows=1 width=8) (actual time=0.027..0.028 rows=1 loops=1)
               Buckets: 524288  Batches: 1  Memory Usage: 4097kB
               ->  Index Scan using idx on foo  (cost=0.00..6.00 rows=1 width=8) (actual time=0.023..0.024 rows=1 loops=1)
                     Index Cond: (b = 2)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 16.278 ms
   (slice0)    Executor memory: 43K bytes.
   (slice1)    Executor memory: 4139K bytes avg x 3 workers, 4163K bytes max (seg0).  Work_mem: 4097K bytes max.
 Memory used:  128000kB
 Execution Time: 119.748 ms
(15 rows)

```

### CODE REVIEW

**Comment by chrishajas** [comment]
Also, can add `set optimizer_trace_fallback=on` to `hash_index.sql` to make sure we're not hitting any asserts/fallbacks, since this looks like it has some tests

File: src/test/regress/expected/create_index_optimizer.out [hunk 1]
```diff
@@ -1337,15 +1337,13 @@ SET maintenance_work_mem = '1MB';
 CREATE INDEX hash_tuplesort_idx ON tenk1 USING hash (stringu1 name_ops) WITH (fillfactor = 10);
 EXPLAIN (COSTS OFF)
 SELECT count(*) FROM tenk1 WHERE stringu1 = 'TVAAAA';
-                       QUERY PLAN                        
+                       QUERY PLAN
 ---------------------------------------------------------
  Finalize Aggregate
    ->  Gather Motion 3:1  (slice1; segments: 3)
          ->  Partial Aggregate
-               ->  Bitmap Heap Scan on tenk1
```
Comment (vraghavan78) [inline src/test/regress/expected/create_index_optimizer.out:1345]: why this change. Also why is Orca not picking a plan here?

Comment (pobbatihari) [reply src/test/regress/expected/create_index_optimizer.out:1345]: Orca able to generate plan for this query, Looks  mistake while copying the change not sure how the pipeline become green. I have adjusted the plan which orca generates. Orca picked the best plan among bitmapscan and index scan.  
# EXPLAIN analyze                                                                                                                                                                        SELECT count(*) FROM tenk1 WHERE stringu1 = 'TVAAAA';
                                                                 QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=0.00..6.01 rows=1 width=8) (actual time=0.938..0.939 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..6.01 rows=1 width=8) (actual time=0.916..0.920 rows=3 loops=1)
         ->  Partial Aggregate  (cost=0.00..6.01 rows=1 width=8) (actual time=0.090..0.090 rows=1 loops=1)
               ->  Index Scan using hash_tuplesort_idx on tenk1  (cost=0.00..6.01 rows=5 width=1) (actual time=0.050..0.074 rows=6 loops=1)
                     Index Cond: (stringu1 = 'TVAAAA'::name)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 9.497 ms
   (slice0)    Executor memory: 43K bytes.
   (slice1)    Executor memory: 45K bytes avg x 3 workers, 45K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 2.041 ms
 
 # set optimizer_enable_indexscan to off;
SET
hmaddileti=# EXPLAIN analyze                                                                                                                                                                        SELECT count(*) FROM tenk1 WHERE stringu1 = 'TVAAAA';
                                                                 QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=0.00..397.99 rows=1 width=8) (actual time=2.259..2.261 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..397.99 rows=1 width=8) (actual time=2.246..2.251 rows=3 loops=1)
         ->  Partial Aggregate  (cost=0.00..397.99 rows=1 width=8) (actual time=1.344..1.345 rows=1 loops=1)
               ->  Bitmap Heap Scan on tenk1  (cost=0.00..397.99 rows=5 width=1) (actual time=1.082..1.337 rows=6 loops=1)
                     Recheck Cond: (stringu1 = 'TVAAAA'::name)
                     ->  Bitmap Index Scan on hash_tuplesort_idx  (cost=0.00..0.00 rows=0 width=0) (actual time=1.063..1.063 rows=6 loops=1)
                           Index Cond: (stringu1 = 'TVAAAA'::name)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 10.422 ms
   (slice0)    Executor memory: 51K bytes.
   (slice1)    Executor memory: 2219K bytes avg x 3 workers, 2219K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 3.514 ms

**Review by chrishajas — COMMENTED** [review]
Do we also need to add logic in `CLogical::PosFromIndex`. I don't think hash has order.

Also, can you add an MDP test since we have some mdp logic that we're adding (`IndexType="Brin"`)

File: src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp [hunk 2]
```diff
@@ -1919,8 +1919,9 @@ CXformUtils::FIndexApplicable(CMemoryPool *mp, const IMDIndex *pmdindex,
 	BOOL possible_ao_table = pmdrel->IsAORowOrColTable() ||
 							 pmdrel->RetrieveRelStorageType() ==
 								 IMDRelation::ErelstorageMixedPartitioned;
-	// GiST can match with either Btree or Bitmap indexes
+	// GiST and Hash can match with either Btree or Bitmap indexes
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:1922]: I don't get it. Is there any test case to covert this?

**Review by gpopt — COMMENTED** [review]
create_index.sql and hash_index.sql are postgres test cases. Can we add test cases into greenplum specific test files?

File: src/test/regress/sql/create_index.sql [hunk 3]
```diff
@@ -396,6 +396,157 @@ SELECT count(*) FROM tenk1 WHERE stringu1 = 'TVAAAA';
 DROP INDEX hash_tuplesort_idx;
 RESET maintenance_work_mem;
 
+--
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/sql/create_index.sql]: sure Chris. Moved testcases to bfv_index.sql. Thank

File: src/test/regress/sql/create_index.sql [hunk 4]
```diff
@@ -396,6 +396,157 @@ SELECT count(*) FROM tenk1 WHERE stringu1 = 'TVAAAA';
...
+-- Test Hash indexes
+--
+
+CREATE TABLE hash_tbl (a int, b int) DISTRIBUTED BY(a);
+INSERT INTO hash_tbl select i,i FROM generate_series(1, 100000)i;
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/sql/create_index.sql]: Yes, Ican test by using various guc's for generating required plan to verify.

File: src/test/regress/sql/create_index.sql [hunk 5]
```diff
@@ -396,6 +396,157 @@ SELECT count(*) FROM tenk1 WHERE stringu1 = 'TVAAAA';
...
+ANALYZE hash_tbl;
+CREATE INDEX gphashind ON hash_tbl USING hash(b);
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/sql/create_index.sql]: Added test cases for hash index on AO table. Yes it choose bitmap scan instead of an index scan.

File: src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp (hunk 2 above)
Comment (pobbatihari) [reply src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:1922]: yes we are covering this, so only we are seeing indexscan operators with hash indexes.  Test cases are covering this change.

File: src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp [hunk 6]
```diff
@@ -1919,8 +1919,9 @@ CXformUtils::FIndexApplicable(CMemoryPool *mp, const IMDIndex *pmdindex,
...
 	if (pmdindex->IndexType() == IMDIndex::EmdindGist ||
+		IMDIndex::EmdindHash == pmdindex->IndexType() ||
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:1924]: reverse the order to be consistent with above line

Comment (pobbatihari) [reply src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp]: done

**Comment by pobbatihari** [comment]
> Do we also need to add logic in `CLogical::PosFromIndex`. I don't think hash has order.
> 
> Also, can you add an MDP test since we have some mdp logic that we're adding (`IndexType="Brin"`)

Added mdp test cases for hash indexes.

File: src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp [hunk 7]
```diff
@@ -1976,6 +1976,7 @@ CPredicateUtils::PexprIndexLookup(CMemoryPool *mp, CMDAccessor *md_accessor,
 	}
 	else if (CUtils::FScalarArrayCmp(pexprScalar) &&
 			 (IMDIndex::EmdindBitmap == pmdindex->IndexType() ||
+			  IMDIndex::EmdindHash == pmdindex->IndexType() ||
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp:1979]: Can we add a test case for this

File: src/test/regress/expected/bfv_index_optimizer.out [hunk 8]
```diff
@@ -703,3 +703,412 @@ RESET enable_bitmapscan;
...
 RESET optimizer_enable_indexscan;
 RESET optimizer_enable_indexonlyscan;
+--
+-- Test Hash indexes
+--
+CREATE TABLE hash_tbl (a int, b int) DISTRIBUTED BY(a);
+INSERT INTO hash_tbl select i,i FROM generate_series(1, 100)i;
+ANALYZE hash_tbl;
+CREATE INDEX hash_idx1 ON hash_tbl USING hash(b);
+-- get non-indexed results for comparison purposes
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/expected/bfv_index_optimizer.out:713]: I mean to check results without setting any guc's. I will reformat the statement.

File: src/test/regress/expected/bfv_index_optimizer.out [hunk 9]
```diff
@@ -703,3 +703,412 @@ RESET enable_bitmapscan;
...
+(9 rows)
+
+DROP INDEX hash_idx2;
+DROP TABLE hash_tbl_ao;
+RESET enable_seqscan;
+RESET optimizer_enable_tablescan;
+-- Test hash indexes with partition table
+CREATE TABLE hash_prt_tbl (a int, b int) DISTRIBUTED BY(a) PARTITION BY RANGE(a)
+(PARTITION p1 START (1) END (500) INCLUSIVE,
+PARTITION p2 START(501) END (1000) INCLUSIVE);
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/expected/bfv_index_optimizer.out:1004]: Currently I am testing with only two partitions p1 and p2. do you want me to check with 5 partitions?

File: src/test/regress/expected/bfv_index_optimizer.out [hunk 10]
```diff
@@ -703,3 +703,412 @@ RESET enable_bitmapscan;
...
+INSERT INTO hash_prt_tbl select i,i FROM generate_series(1, 1000)i;
+ANALYZE hash_prt_tbl;
+CREATE INDEX hash_idx3 ON hash_prt_tbl USING hash(b);
+-- get non-indexed results for comparison purposes
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/expected/bfv_index_optimizer.out:1008]: ok

File: src/test/regress/expected/bfv_index_optimizer.out [hunk 11]
```diff
@@ -703,3 +703,412 @@ RESET enable_bitmapscan;
...
+(4 rows)
+
+SELECT * FROM hash_tbl WHERE b=3 or b=5;
+ a | b 
+---+---
+ 5 | 5
+ 3 | 3
+(2 rows)
+
+-- Now check the results from seqscan
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/expected/bfv_index_optimizer.out:764]: I do agree. I will remove seqscan results.

File: src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp [hunk 12]
```diff
@@ -2432,7 +2433,8 @@ CXformUtils::PexprBuildBtreeIndexPlan(
 	IMdIdArray *partition_mdids = nullptr;
 
 	if (ptabdesc->RetrieveRelStorageType() != IMDRelation::ErelstorageHeap &&
-		pmdindex->IndexType() == IMDIndex::EmdindGist)
+		(pmdindex->IndexType() == IMDIndex::EmdindGist ||
+		 pmdindex->IndexType() == IMDIndex::EmdindHash))
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:2437]: This change is for to avoid indexscan for the  AO tables.   Yes testcases covering this scenario.

**Comment by pobbatihari** [comment]
> create_index.sql and hash_index.sql are postgres test cases. Can we add test cases into greenplum specific test files?

yup moved testcases under bfv_index.sql

File: src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp (hunk 7 above)
Comment (pobbatihari) [reply src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp]: As discussed, this check doesn't look to be correct for hash indexes here. Removing

**Review by gpopt — APPROVED** [review]
+1

**Review by chrishajas — APPROVED** [review]
Looks good, awesome work!
